### PR TITLE
Fix/public ip arg build ipv6

### PR DESCRIPTION
**Pull request by kshokhin**

### CODE REVIEW

File: app/teovpnstarter.c [hunk 1]
```diff
@@ -10,18 +10,162 @@ static const char* home_env_var = "HOME";
...
+        return 0;
+    }
+
+    char *ipv6_url_request_buffer = NULL;
+    const char* ipv6_url_pattern = "http://169.254.169.254/latest/meta-data/network/interfaces/macs/ipv6s";
+    //NOTE: buffer holds string like 02:43:8a:02:50:57/, so it already has '/'
+    size_t ipv6_url_buffer_size = strlen(ipv6_url_pattern) + strlen(buffer) + 1;
+    ipv6_url_request_buffer = (char*)malloc(ipv6_url_buffer_size);//http://169.254.169.254/latest/meta-data/network/interfaces/macs/mac-address/ipv6s
+    snprintf(ipv6_url_request_buffer, ipv6_url_buffer_size,
+        "http://169.254.169.254/latest/meta-data/network/interfaces/macs/%sipv6s", buffer);
```
Comment (chernikovdmitry) [inline app/teovpnstarter.c:77]: Config?
```suggestion
    snprintf(ipv6_url_request_buffer, ipv6_url_buffer_size, 
        "%s%sipv6s", ipv6_request_url, buffer);
```

File: app/teovpnstarter.c [hunk 2]
```diff
@@ -10,18 +10,162 @@ static const char* home_env_var = "HOME";
...
+    /* Perform the request, res will get the return code */
+    CURLcode res = curl_easy_perform(curl);
+
+    if (res != CURLE_OK) {
+        printf("request macs failed: %s\n", curl_easy_strerror(res));
+        return 0;
+    }
+
+    char *ipv6_url_request_buffer = NULL;
+    const char* ipv6_url_pattern = "http://169.254.169.254/latest/meta-data/network/interfaces/macs/ipv6s";
```
Comment (chernikovdmitry) [inline app/teovpnstarter.c:72]: Config?

File: app/teovpnstarter.c [hunk 3]
```diff
@@ -10,18 +10,162 @@ static const char* home_env_var = "HOME";
...
+        return 0;
+    }
+
+    long response_code = 0;
+    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
+
+    printf("%s\n%ld\n", buffer, response_code);
+
+    if (response_code == 200) {
+        settings->public_ipv6 = strdup(buffer);
```
Comment (chernikovdmitry) [inline app/teovpnstarter.c:101]: same, no need in strdup

**Review by chernikovdmitry — CHANGES_REQUESTED** [review]
Looks like it would leak memory

File: app/teovpnstarter.c [hunk 4]
```diff
@@ -10,18 +10,234 @@ static const char* home_env_var = "HOME";
...
+        set_buffer(buffer, data, data_size);
+    } else {
+        size_t old_capacity = buffer->capacity;
+        while(buffer->capacity - buffer->size < data_size) {
+            buffer->capacity *= 2;
+        }
+
+        if (old_capacity != buffer->capacity) {
+            //TOOD: realloc can fail
+            buffer->data = (char*)realloc(buffer->data, buffer->capacity);
```
Comment (angelskieglazki) [inline app/teovpnstarter.c:79]: ccl_realloc can use

**Review by angelskieglazki — APPROVED** [review]

File: app/teovpnstarter.c [hunk 5]
```diff
@@ -14,23 +14,90 @@ typedef struct AppSettings
...
+void set_buffer(Buffer *buffer, const char* data, size_t data_size) {
+    if (buffer == NULL || data == NULL || data_size == 0) { return; }
+
+    free(buffer->data);
+
+    buffer->data = (char*)malloc(data_size);
+    memcpy(buffer->data, data, data_size);
+    buffer->size = data_size;
+    buffer->capacity = data_size;
+}
```
Comment (chernikovdmitry) [inline app/teovpnstarter.c:64]:
```suggestion
void set_buffer(Buffer *buffer, const char* data, size_t data_size) {
    if (buffer == NULL) { return; }

    free(buffer->data);

    if (data == NULL || data_size == 0) {
      buffer->data = NULL;
      buffer->capacity = 0;
      buffer->size = 0;
      return;
    }

    buffer->data = (char*)malloc(data_size);
    memcpy(buffer->data, data, data_size);
    buffer->size = data_size;
    buffer->capacity = data_size;
}
```

File: app/teovpnstarter.c [hunk 6]
```diff
@@ -14,23 +14,90 @@ typedef struct AppSettings
...
-    char *buffer = NULL;
+    Buffer data_buffer = create_buffer();
 
     curl_easy_setopt(curl, CURLOPT_URL, "http://169.254.169.254/latest/meta-data/public-ipv4");
     curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, save_data_cb);
-    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &buffer);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data_buffer);
 
     /* Perform the request, res will get the return code */
     CURLcode res = curl_easy_perform(curl);
```
Comment (chernikovdmitry) [inline app/teovpnstarter.c:103]: At lines 105-108 should be
```
    if (res != CURLE_OK) {
        printf("request_public_ipv4 failed: %s\n", curl_easy_strerror(res));
        destroy_buffer(&data_buffer);
        return 0;
    }
```

File: app/teovpnstarter.c [hunk 7]
```diff
@@ -44,21 +111,22 @@ int request_public_ipv4(CURL *curl, AppSettings *settings) {
...
 int request_public_ipv6(CURL *curl, AppSettings *settings) {
-    char *buffer = NULL;
+    Buffer data_buffer = create_buffer();
     curl_easy_setopt(curl, CURLOPT_URL, "http://169.254.169.254/latest/meta-data/network/interfaces/macs/");
     curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, save_data_cb);
-    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &buffer);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data_buffer);
 
     /* Perform the request, res will get the return code */
     CURLcode res = curl_easy_perform(curl);
```
Comment (chernikovdmitry) [inline app/teovpnstarter.c:132]: Same, destroy_buffer(&data_buffer); if (res != CURLE_OK)

File: app/teovpnstarter.c [hunk 8]
```diff
@@ -68,22 +136,23 @@ int request_public_ipv6(CURL *curl, AppSettings *settings) {
...
 
     printf("%s\n", ipv6_url_request_buffer);
 
     curl_easy_setopt(curl, CURLOPT_URL, ipv6_url_request_buffer);
     curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, save_data_cb);
-    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &buffer);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data_buffer);
 
     res = curl_easy_perform(curl);
     if (res != CURLE_OK) {
```
Comment (chernikovdmitry) [inline app/teovpnstarter.c:158]: destroy_buffer(&data_buffer);

**Review by chernikovdmitry — CHANGES_REQUESTED** [review]

**Review by chernikovdmitry — APPROVED** [review]
Cool story :cool:

**Pull request merged**